### PR TITLE
WIP: Specific calls

### PR DESCRIPTION
**Pull request by jbcaillau**
@ocots I have created this new branch from the `variable_constraints` one, where I am implementing the use of one `variable` (see #14). This has the side effect to more or less deprecate the "specific calls" you added (for compatibility with what @PierreMartinon and @joseph-gergaud) did for direct solving (scalar / vector issue) involving the `_Time` type in `functions.jl`. I am afraid of having made big 💩💩 as right now, to pass the tests (see new tests in `test_model.jl`) like

I have
- updated `BoundaryConstraint` and `Mayer` wrapper (part of the code / comments are probably obsolete, now)
- commented the `functions` test in `runtests.jl`

Could you please have a look?

In between, I am preparing something clean in `model.jl` (essentially `nlp_constraints` function) so that @PierreMartinon and @joseph-gergaud can update the optimisation problem build.

### CODE REVIEW

**Comment by ocots** [comment]
I have refactored functions and model. The tests should pass and I think we can merge into develop.

**Comment by jbcaillau** [comment]
@ocots very nice job on the branch. I have noted that you still need to add some tests. I merged into `constraints` (WIP), it is OK. 

After reading some stuff on parametric types (subtle covariance / invariance things), I feel like we should use types to parametrise instead of literal values (such as `Symbols` - even if types are just special values). For instance, an `OptimalControlProblem` can indeed be autonomous or not, and it is nice that `OptimalControlProblem{Autonomous}` be a subtype of `OptimalControlProblem`. Hence, if suggest a (hopefully) little 😬 change according to:
```julia
abstract type TimeDependence end
abstract type Autonomous <: TimeDependence
abstract type NonAutonomous <: TimeDependence

abstract type VariableDependence end
abstract type Variable <: VariableDependence
abstract type NonVariable <: VariableDependence

struct OptimalControlProblem{T <: TimeDependence, V <: VariableDependence}
    ...
end
```
This allows, *e.g.*, to define functions on `OptimalControlProblem` when they equally work autonomous / non autonomous / variable / non variable problems, `OptimalControlProblem{TimeDependence, Variable}` for those who don't care about autonomous / non autonomous but do care about having a variable or not, *etc.*

What do you think?

**Comment by ocots** [comment]
Several remarks:

- Actually, I have finished to update `functions` and `model`: code and tests. 

- I have added some tests but we could had so many more.

- There are some changes due to the fact that we want to authorize calls with an empty vector (here `v = [ ]`):

https://github.com/control-toolbox/CTBase.jl/blob/5a8c4f67f0a92296529a34ac5ffed339550e986e/test/test_model.jl#L759

recalling that ξ is for `ControlConstraint`. I have introduced in the case of no variable dependence, the following signature:

https://github.com/control-toolbox/CTBase.jl/blob/5a8c4f67f0a92296529a34ac5ffed339550e986e/src/function.jl#L272

where 

https://github.com/control-toolbox/CTBase.jl/blob/5a8c4f67f0a92296529a34ac5ffed339550e986e/src/CTBase.jl#L70

I have added this because `[ ]` is not a `AbstractVector{<:Real}` but `AbstractVector{Any}`.

- But after that, I have noticed that we have the same problem for `BoundaryConstraint`. If a user defines for instance an initial constraint:

https://github.com/control-toolbox/CTBase.jl/blob/5a8c4f67f0a92296529a34ac5ffed339550e986e/test/test_model.jl#L442

and then get it from the `constraint` function, this constraint has been cast and so the user has to provide `x0` but also `xf`: 

https://github.com/control-toolbox/CTBase.jl/blob/5a8c4f67f0a92296529a34ac5ffed339550e986e/test/test_model.jl#L438

Here, I pass `[ ]` for `xf` but I don't know if it is ok to let this possibility. To make thinks simple I have changed the alias `ctVector`, from:

https://github.com/control-toolbox/CTBase.jl/blob/5a8c4f67f0a92296529a34ac5ffed339550e986e/src/CTBase.jl#L37

to

https://github.com/control-toolbox/CTBase.jl/blob/5a8c4f67f0a92296529a34ac5ffed339550e986e/src/CTBase.jl#L38.

**Comment by jbcaillau** [comment]
@ocots 👍🏽having a look at what you did.

On `[ ]`: I assume it is due to the fact that, after casting the functions, whatever the type of `OptimalControlModel`, we want to be able to call, *e.g.* a state constraint, using the signature
```julia
(t::Time, x::State, v::Variable)
```
But we can actually cast `[ ]` to be a `ctVector` (that is a `Variable`, or so):
```
julia
julia> ctVector = Union{Real, Vector{<:Real}} # useless to define ctNumber = Real?
Union{Real, Vector{<:Real}}

julia> v = Vector{Real}()
Real[]

julia> v isa ctVector
true
```

**Comment by ocots** [comment]
Are you ok to make a pass on the code?

**Comment by ocots** [comment]
Do you make a pass on the code or do you prefer I do it? 

What do you prefer for the user to declare an optimal control model:

By the type:

```julia
ocp = OptimalControlModel{Autonomous, NonVariable}()
```

By a constructor, assuming that `(Autonomous, NonVariable)` are the default values and that the order for the dependences is not important. All lines below are equivalent:

```julia
ocp = Model()
ocp = Model(Autonomous)
ocp = Model(NonVariable)
ocp = Model(Autonomous, NonVariable)
ocp = Model(NonVariable, Autonomous)
```

Or something else: 

Default values:
- [ ] yes
- [ ] no

Keywords arguments:
- [ ] yes
- [ ] no

Fix an order in the signature for the dependences:
- [ ] yes
- [ ] no

To not fix the order, we can simply define:

```julia
function Model(dep::Union{TimeDependence, VariableDependence}...)::OptimalControlModel{TimeDependence, VariableDependence}
    # some checkings: 
    # length(dep)≤2, at most one TimeDependence and one VariableDependence
    td = Autonomous ∈ dep ? Autonomous : NonAutonomous
    vd = NonVariable ∈ dep ? NonVariable : Variable
    return OptimalControlModel{td, vd}()
end
```

**Comment by jbcaillau** [comment]
@ocots I haven't touch the code. Yes, the kind of constructor you wrote seems well suited for the user: boolean keyword args with default to autonomous and no variable. 
```julia
Model(; autonomous=true, variable=false)::OptimalControlProblem(TimeDependence, VariableDependence)
```
NB. `Autonomous` ... `NonVariable` are subtypes, not values of `TimeDependence`, *etc.* So
```julia
julia> Autonomous <: TimeDependence
true
Autonomous in TimeDependence
false
```

**Comment by ocots** [comment]
Yes my constructor is wrong it is still possible to pass types instead of booleans.

**Comment by ocots** [comment]
Need more work to be done. I close the PR and we will reopen it later.

**Comment by jbcaillau** [comment]
@ocots OK. Trying to catch up with you on phone this afternoon if possible. As this is a WIP PR, I reopen it (I though it was merged!)

File: src/model.jl [hunk 1]
```diff
@@ -416,87 +393,240 @@ julia> ocp.time_name
...
+        !all(1 .≤ rg .≤ n) && throw(IncorrectArgument("the range $rg of the final state constraint must be contained in 1:$n"))
+    elseif type == :variable
+        !all(1 .≤ rg .≤ q) && throw(IncorrectArgument("the range $rg of the variable constraint must be contained in 1:$q"))
+    end
+
     # set the constraint
-    if type ∈ [ :initial, :final ] # not allowed for :control or :state (does not make sense)
-        ocp.constraints[label] = (type, :eq, x -> x, val, val)
+    if type == :initial # not allowed for :control or :state (does not make sense)
+        B = nothing
```
Comment (jbcaillau) [inline src/model.jl:460]: Why not
```julia
BoundaryConstraint(vd == :v_indep ? (x0, xf) -> x0[rg] : (x0, xf, v) -> x0[rg])
```

NB. Actually, one could also dispatch on the ocp type.

**Comment by ocots** [comment]
@BaptisteCbl En fait on peut faire

```julia
time_dep(ocp::OptimalControlProblem{NonAutonomous, <:VariableDependence}) = true
```

**Comment by jbcaillau** [comment]
@ocots I have corrected the examples and added subtyping in https://github.com/control-toolbox/CTBase.jl/pull/59#issuecomment-1537143678 above.

**Comment by jbcaillau** [comment]
IMG_2091

**Comment by jbcaillau** [comment]
@ocots On [this](https://github.com/control-toolbox/CTBase.jl/pull/59#issuecomment-1537143678), do we agree that with an appropriate cast there is no need for the extra type `EmptyDecisionVariable`?
IMG_2092

**Comment by ocots** [comment]
Yep.

I have change Variable to DecisionVariable you have defined a type Variable for the dependence. Maybe it would have been better to do the contrary.

**Comment by jbcaillau** [comment]
@ocots @BaptisteCbl On "value types", check [the doc](https://docs.julialang.org/en/v1/manual/types/#%22Value-types%22)

**Pull request merged**